rust_src/src/frames.rs

```diff
@@ -77,4 +91,68 @@ pub fn selected_frame() -> LispObject {
     unsafe { LispObject::from(current_frame) }
 }
 
+/// Return non-nil if OBJECT is a frame which has not been deleted.
+/// Value is nil if OBJECT is not a live frame.  If object is a live
+/// frame, the return value indicates what sort of terminal device it is
+/// displayed on.  See the documentation of `framep' for possible
+/// return values.
+#[lisp_fn]
+fn frame_live_p(object: LispObject) -> LispObject {
+    if object.as_frame().map_or(false, |f| f.is_live()) {
+        LispObject::from(framep(object))
+    } else {
+        LispObject::constant_nil()
+    }
+}
+
+/// Set selected window of FRAME to WINDOW.
+/// FRAME must be a live frame and defaults to the selected one.  If FRAME
+/// is the selected frame, this makes WINDOW the selected window.  Optional
+/// argument NORECORD non-nil means to neither change the order of recently
+/// selected windows nor the buffer list.  WINDOW must denote a live window.
+/// Return WINDOW.
+#[lisp_fn(min = "2")]
+pub fn set_frame_selected_window(
+    frame: LispObject,
+    window: LispObject,
+    norecord: LispObject,
+) -> LispObject {
+    let mut frame_ref = frame_live_or_selected(frame);
+    let w = window.as_live_window_or_error();
+
+    if frame_ref != w.frame().as_frame().unwrap() {
+        error!("In `set-frame-selected-window', WINDOW is not on FRAME")
+    }
+    if frame_ref == selected_frame().as_frame().unwrap() {
+        unsafe { LispObject::from(Fselect_window(window.to_raw(), norecord.to_raw())) }
+    } else {
```

Comment:
I have been pondering updating the API because the `as_foo` methods are almost always returning a foo so the unwrap is just noise. However, we clearly also need the Option return version as well.

Reply:
Yeah I also think that's annoying, but I couldn't think of a better solution.

Reply:
One possibility is moving to `Result` returns for lisp_fn. Then we could use `?` to simplify.

`foo.as_string()?` this way we can handle it locally or let the error bubble out. This would also reduce the need for `as_foo_or_error`

Reply:
Result returns might be coming in place of `error!()` at some point but are a long way off.

In any case, wherever we're currently using `unwrap()` would stay that way even with Results, since it marks places where internal invariants dictate presence of `Some`. A controlled abort is the correct way to handle a violation.